### PR TITLE
More stuff

### PR DESCRIPTION
**Pull request by Hellowlol**
So this PR got alot bigger then what i indended.

Anyways this adds support for more controls and make this easier to work with. Everything is a class and all the props are hardcoded as i didnt like the default name that tesla uses or it returns some kind of info that isnt very user friendly. (center_display_state value 8 == dog mode) etc.

This is a wip, need some cleanup, add more logging and tests.

### CODE REVIEW

**Comment by Dreamsorcerer** [comment]
That is big. :P
I'll try and take an initial look through in a few days. Note that it will be easier if you don't make a lot of formatting changes (maybe you were running Black or something automatically?).

**Comment by Hellowlol** [comment]
Yeah, its black. It runs each time I hit save 🤦‍♂️. I’ll disable it now. 

I’ll add some tests and examples on why I think this Is a better solution. The goal is to have a user friendly python classes while still able to use the raw json response if needed.

I’m not sure how to add the test as I have never used azure pipelines before, but I’ll figure that out.

Hopefully I’ll  have this ready over the weekend, I just wanted to get some kinda feedback. 


Thanks!

**Comment by Dreamsorcerer** [comment]
I think @mlowijs was experimenting with the Azure Pipelines, but I don't think he finished setting it up.

So, just ignore that. As long as I can run the tests locally, that should be enough. Though feel free to setup CI integration with Travis or something similar.

File: tesla_api/misc.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+
+def c_to_f(value):
+    return value * 1.8 + 32
+
+
+def cast(value):
+    """Helper"""
+    if isinstance(value, str):
+        return False if value == "0" else True
```
Comment (Dreamsorcerer) [inline tesla_api/misc.py]:
```suggestion
        return value != "0"
```

File: tesla_api/misc.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    if isinstance(value, int):
+        return False if value is 0 else True
```
Comment (Dreamsorcerer) [inline tesla_api/misc.py]:
```suggestion
        return value != 0
```
Don't use `is`. It`ll probably work, but it's not guaranteed (it's implementation dependant).

File: tesla_api/speedlimit.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    @property
+    def max_limit_mph(self):
+        return self._vehicle._data["vehicle_state"]["speed_limit_mode"]["max_limit_mph"]
+
+    def max_limit(self):
+        """Max limit in the gui format"""
+        return self._vehicle._format_distance_unit(self.max_limit_mph)
+
+    @property
+    def pin_clode_set(self):
```
Comment (Dreamsorcerer) [inline tesla_api/speedlimit.py]:
```suggestion
    def pin_code_set(self):
```

**Comment by Dreamsorcerer** [comment]
I'll try and take a broader look over the weekend and see if I can get an understanding of it.

Note that this may intersect with my changes in https://github.com/mlowijs/tesla_api/pull/33/files, which I started sometime ago, but haven't got around to finishing off. That PR adds very strict typing throughout the full code base, enforced with Mypy.

**Comment by Hellowlol** [comment]
Thats nice, i plan to use typing, i started on the low hanging fruit. But i got bored. I'll run monkeytype when after all the tests are done.

**Comment by Hellowlol** [comment]
Ok, this **_beast_** should the ready for renew, it isn't done but I suspect it will take some time to do the review.

The goal of the PR is to have classes that hides the logic for what value to find and how to format it.
Fx is the speedlimit requires the limit in mile per hour is will check what unit type the car uses and convert if needed. Same goes for other attributes where the api returns the data in another format then what the car uses.

Raw json response supported via ``` vehicle.get_charge_state``` etc

Removed:
 - azure_piplines (seemed to be broken anyway)

Added/Improved:
- Github action is now used for CI
- 75 test written (and counting)
- 86 % coverage, i want to cover everything not energy related. https://codecov.io/gh/Hellowlol/tesla_api/branch/more-controls
- package upload

TODO:
Readme will updated when the PR has been approved.
Update the doc string to better explain what this property mean/does

And down the road:
- lint with black
- Add typing and lint with mypy
- Add docs to readthedocs

**Comment by Dreamsorcerer** [comment]
OK, I'll take a look this week. Let's see if we can break it up into some smaller pieces as we go.

**Comment by Dreamsorcerer** [comment]
Don't bother with conflicts right now. I'm going to merge the Mypy stuff first, which will just create more conflicts.

**Comment by Hellowlol** [comment]
Ok. Ill still add the typing as I need this to create the docs. Im to lazy to write them myself.

Pretty awsome how much its possible to semi automate. Use a tool that creates the tests that creates typing hints that is used to create the docs 😂

**Comment by Hellowlol** [comment]
Closing this as the progress is too slow. Ill maintain my own fork. Thanks anyway. 😃

**Pull request closed without merging**